packages/nx/src/plugins/js/project-graph/build-dependencies/target-project-locator.ts

```diff
@@ -556,14 +569,29 @@ export class TargetProjectLocator {
 
       while (dir !== dirname(dir)) {
         const packageJsonPath = join(dir, 'package.json');
-        try {
-          const parsedPackageJson = readJsonFile(packageJsonPath);
-          // Ensure the package.json contains the "name" and "version" fields
-          if (parsedPackageJson.name && parsedPackageJson.version) {
-            return parsedPackageJson;
+        if (this.packageJsonResolutionCache.has(packageJsonPath)) {
+          return this.packageJsonResolutionCache.get(packageJsonPath);
+        }
+        if (existsSync(packageJsonPath)) {
+          try {
+            const parsedPackageJson = readJsonFile(packageJsonPath);
+            // Ensure the package.json contains the "name" and "version" fields
+            if (parsedPackageJson.name && parsedPackageJson.version) {
+              this.packageJsonResolutionCache.set(
+                packageJsonPath,
+                parsedPackageJson
+              );
+              return parsedPackageJson;
+            } else {
+              this.packageJsonResolutionCache.set(packageJsonPath, null);
+              return null;
+            }
+          } catch {
+            // Package.json is invalid, keep traversing
           }
-        } catch {
-          // Package.json doesn't exist, keep traversing
+        }
+        if (dir === workspaceRoot) {
+          return null;
```

Comment:
So before.. we were able to.. resolve outside of the workspace root?

Have we considered if this actually helps when the workspaceRoot is not the same as the node_modules root? WHat if someone has like..

repo-root/package.json
repo-root/node_modules
repo-root/nx-workspace-root/nx.json

If this doesn't yield a major perf behefit, I would not make htis change for now